start-workflow-wfh allows to pass mediapackage id from previous WFH

ts23 opened the pull request:
Update StartWF WFH to allow passing MP ids in from previous WF operations. this can be useful for Duplicate MP workflows that can then start other operation on the duplicates

mtneug commented:
@ts23 why not simply use `<configuration key="media-package">${duplicate_media_package_1_id}</configuration>`?

ts23 commented:
> @ts23 why not simply use `<configuration key="media-package">${duplicate_media_package_1_id}</configuration>`?

doesn't solve the issue of multiple mps being created

mtneug commented:
Alternative approach to keep it consistent with other operations:

- extend the `duplicate-event` operation with an additional variable `duplicate_media_package_ids` that holds a comma-separated list of the new media package ids
- change the `media-package` property of the `start-workflow` operation such that multiple comma-separated media package ids can be supplied similar to how the common `source-flavors` property works
- it might be a good idea to change/deprecate the property name from `media-package` to `media-packages` (again `source-flavor` -> `source-flavors` serves as a precedent)

ts23 committed:
Update Start Workflow WFH to take comma separated list of mediapackages
Update Duplicate Workflow WFH to provide comma separated list of mediapackages
Modify the Duplicate Event Workflow to demonstrate Duplicate WFH and StartWorkflow WFH in combination

# Conflicts:
#	docs/guides/admin/docs/workflowoperationhandlers/start-workflow-woh.md
#	etc/workflows/duplicate-event.xml
#	modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java

mtneug commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java:
```diff
@@ -104,22 +97,33 @@ public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobConte
+      // Get media package
+      Opt<MediaPackage> mpOpt = assetManager.getMediaPackage(mpId);
+      if (mpOpt.isNone()) {
+        throw new WorkflowOperationException(format("Media package %s not found", mpId));
```
Maybe it would be better to record the error but continue with the other IDs and only then fail? Or do you think a fast fail would be better?

ts23 replied:
we should make that optional on fail_on_error (fast fail if fail-on-error = true, record and continue if fail-on-error=false)

mtneug commented on etc/workflows/duplicate-event.xml:
```diff
@@ -59,6 +59,14 @@
+        <configuration key="workflow_definition">fast</configuration>
+      </configurations>
+    </operation>
+
```
I assume this is only for the review? Could you remove it for the merge?

ts23 replied:
I was going to leave the example in but it be default. Or do we not provide usage examples?

mtneug replied:
I would remove it as this is most likely used in production and not everybody will use `fast` for that. In any case, `workflow_definition` should be `workflow-definition`.

mtneug reviewed (changes requested):
The feature works and only have some small points.

gregorydlogan commented:
Does this related to #3155?

mtneug commented:
No this is for the `start-workflow` operation. #3155 allows access to organisation properties in workflows.

mtneug commented on modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java:
```diff
@@ -51,7 +53,7 @@
 
   /** Name of the configuration option that provides the media package ID */
-  public static final String MEDIA_PACKAGE_ID = "media-package";
+  public static final String MEDIA_PACKAGE_ID = "media-packages";
```
This is a breaking change that is not allowed in OC 11. I would suggest only deprecating the `media-package` option but still allowing it. It should be possible to simply merge the `media-packages` and `media-package` strings possible with a seperator.

ts23 replied:
should be fixed

mtneug reviewed (changes requested):
This works, but there are still two remaining issues I found. I'm sorry for the late review.

ts23 committed:
Update Start Workflow WFH to allow media-package and media-packages

# Conflicts:
#	etc/workflows/duplicate-event.xml

ts23 committed:
Update Start Workflow WFH to consider fail-on-error for fast fail or collect errors

# Conflicts:
#	modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java

mtneug commented on docs/guides/admin/docs/workflowoperationhandlers/start-workflow-woh.md:
```diff
@@ -20,7 +20,7 @@ The StartWorkflowWorkflowOperationHandler can be used to start a new workflow fo
   <configurations>
     <configuration key="workflow-definition">fast</configuration>
-    <configuration key="media-package">e72f2265-472a-49ae-bc04-8301d94b4b1a</configuration>
+    <configuration key="media-packages">${duplicate\_media\_package\_ids}</configuration>
```
`_` in backticks does not need to be escaped:

```suggestion
    <configuration key="media-packages">${duplicate_media_package_ids}</configuration>
```

mtneug commented on docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md:
```diff
@@ -16,7 +16,10 @@ For each duplicated event the new media package ID is stored as a workflow prope
 |Name                                    |Example                                                             |Description                                    |
 |----------------------------------------|--------------------------------------------------------------------|-----------------------------------------------|
-|duplicate\_media\_package\_*number*\_id |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       |
+|duplicate\_media\_package\_*number*\_id |`duplicate\_media\_package\_1\_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
+|duplicate\_media\_package\_ids          |`duplicate\_media\_package\_ids=e72f2265-472a-49ae-bc04-8301d94b4b1a, a32e2265-472a-49ae-bc04-8351d94b4b1c` | comma separated list of Media package IDs of the duplicated event |
```
`_` in backticks does not need to be escaped:

```suggestion
|duplicate\_media\_package\_*number*\_id |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
|duplicate\_media\_package\_ids          |`duplicate_media_package_ids=e72f2265-472a-49ae-bc04-8301d94b4b1a, a32e2265-472a-49ae-bc04-8351d94b4b1c` | comma separated list of Media package IDs of the duplicated event |
```

mtneug reviewed (changes requested):
Only some nitpicking in the documentation. Can go in otherwise.

Thanks for the patience.

ts23 committed:
Update docs/guides/admin/docs/workflowoperationhandlers/start-workflow-woh.md

Co-authored-by: Matthias Neugebauer <mtneug@mailbox.org>

ts23 committed:
Update docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md

Co-authored-by: Matthias Neugebauer <mtneug@mailbox.org>

KatrinIhler commented on etc/workflows/duplicate-event.xml:
```diff
@@ -59,6 +59,14 @@
+      id="start-workflow">
+      <configurations>
+        <configuration key="media-packages">${duplicate_media_package_ids}</configuration>
+        <configuration key="workflow-definition">fast</configuration>
```
The WOH itself is cool, but I don't agree with this change to the workflow definition. We shouldn't include fast, since it's a testing workflow and people shouldn't run this by accident on their production cluster (as just happened to us), and also - why include an import workflow here? Since the initial event that was duplicated was already processed, the import has already happened, the only thing we might want to do is publish - but I wouldn't do that by default either.
